### DISCUSSION TITLE
Transcribing live stream calls from Twilio in python

### DISCUSSION THREAD

**Discussion by dquang-tran** [deepgram/community#311] [General help] [answered]
Hi team,

Love the library — amazing job!

I am currently attempting to implement live streaming audio transcription.
This article gives an example of transcribing from live streaming audio: https://developers.deepgram.com/docs/getting-started-with-live-streaming-audio

I would like to implement something similar but with Twilio audio as the source. However, I'm currently running into the issue where TRANSCRIPT_RECEIVED event handler doesn't trigger: https://replit.com/join/tfgzvbjuws-quangtran73

Curious if I'm doing something wrong and whether there are tips to making this work?

**Comment by nikolawhallon** [comment, marked as answer]
This is a guide for streaming live Twilio calls to Deepgram in Python: https://deepgram.com/learn/deepgram-twilio-streaming

It uses Twilio's streaming API, which will fork Twilio audio to your server, which then processes the audio, sends it to Deepgram via websockets, and gets the live streaming transcription results. Can you share more about what you have tried? My first guess is that, Twilio media messages can't be directly parsed by Deepgram, they are streamed as json text, where the audio payloads are base64-encoded mulaw, so some processing of Twilio's media messages needs to happen before sending the audio to Deepgram (that guide should cover this).

**Reply by dquang-tran** [reply]
Hi @nikolawhallon,

The media messages from Twilio in my replit are processed the same way as that guide's [sample code](https://github.com/deepgram-devs/deepgram-twilio-streaming-python/blob/master/twilio.py). I've tried adjusting my code to process the buffer through AudioSegment as well. 

One thing that is different is that my code uses the python sdk, specifically the deepgram.transcription.live object for live streaming.
Looking at the dashboard logs, the transcriptions are being received but the audio length is 0.
<img width="693" alt="Screenshot 2023-08-09 at 09 41 40" src="https://github.com/deepgram/community/assets/22584558/36fb54a7-4935-4f7c-912e-df901170eab7">

**Reply by nikolawhallon** [reply]
I see - sorry, had to signup for replit to take a look at the code! In general, when live streaming and Deepgram is receiving no audio, or unexpected audio, the best place to start troubleshooting is here: https://github.com/deepgram/streaming-test-suite#validate-your-audio

But I can see a few things in the code which stand out. You configure Deepgram via:
```
    deepgramLive = await deepgram.transcription.live({
      'smart_format': True,
      'interim_results': False,
      'language': 'en-US',
    })
```
Since Twilio sends raw mulaw audio, you'll need to add `encoding=mulaw` and `sample_rate=8000` to this list - also `channels=1` or `channels=2&multichannel=true` if handling multiple channels (based on the audio processing logic, this seems like you expect Twilio to stream in just one channel of audio, which is the default configuration for Twilio streams).

Without specifying those parameters, Deepgram assumes the incoming audio is containerized, and if it isn't, then Deepgram will consider the audio as corrupt (it won't know how to process the audio).

The other thing I notice is:
```
      deepgramLive.send(buffer)
      buffer = bytearray(b'')
      deepgramLive.send({"type": "KeepAlive"})
```
Sending a KeepAlive message here isn't necessary.

If these don't resolve the issue, using the streaming test suite to inspect and validate the audio you are passing through could be very helpful.

**Reply by nikolawhallon** [reply]
One more thing I will note, you have:
```
      if chunk == b'':
        empty_byte_received = True
```
I would ignore this case - Twilio doesn't use the same convention of "empty byte = close the connection" that Deepgram does, so this might be prematurely closing the connection by accident when you later forward this to Deepgram.

**Reply by dquang-tran** [reply]
Thanks! Adding sample rate and encoding configs made it so that I now receive TRANSCRIPT_RECEIVED events through my app.
However, another problem I'm now running into is that deepgram stops emitting them after some number of events. In addition, the transcriptions received contain no words :(

**Reply by nikolawhallon** [reply]
What does the audio duration say in Console? Do you have a request ID? I can take a look.

**Reply by dquang-tran** [reply]
Removing empty byte logic helped with deepgram not prematurely closing the connection!
Re words list being empty, the request ID is ce1b9595-4e0d-468f-a6d7-c9438f594966. The audio duration is as expected and non 0.
Additionally, is it necessary to process data with AudioSegment before sending it?
@nikolawhallon

**Reply by nikolawhallon** [reply]
Not necessary, no, I think some of our example code uses AudioSegment just to help stitch together the inbound and outbound streams into a single stereo stream.

**Reply by nikolawhallon** [reply]
I looked at that request, and also noticed the code changed a bit on the replit. Can you try using this block:

```
    if len(buffer) >= BUFFER_SIZE:
      deepgramLive.send(buffer)
      buffer = bytearray(b'')
```

and then can you try enforcing the track, e.g.

```
    if data['event'] == "media":
      media = data["media"]
      if media["track"] == "inbound":
```

Edit: it would also be useful to ensure that `media['timestamp']` is constantly increasing by 20ms

**Reply by dquang-tran** [reply]
That did help! Now I'm getting non empty transcripts. However, the transcription is gibberish — not corresponding to audio input at all :/

Saw that your [code](https://github.com/deepgram-devs/deepgram-twilio-streaming-python/blob/master/twilio.py#L20) uses asyncio queue with synchronous logic to send data to deepgram. Could it be the issue that the sound data is out of order since I'm just sending data with `deepgramLive.send(buffer)`?

>it would also be useful to ensure that media['timestamp'] is constantly increasing by 20ms

Can you elaborate more on that? Verified that this is indeed the case by so don't think this is what's causing an issue.

**Reply by nikolawhallon** [reply]
> However, the transcription is gibberish — not corresponding to audio input at all

Can you share a request id at this point too? If the transcript is gibberish, it's likely the audio is still gibberish/corrupt/not formatted well for some reason, see the below point too:

>  Can you elaborate more on that? Verified that this is indeed the case by so don't think this is what's causing an issue.

Twilio can drop packets, and when it does audio will come in with timestamps that seem to skip, like 0, 20, 40, 80, 100 (notice the 60-80ms audio was skipped) - when this happens, simply appending the audio ends up in very strange audio where words can kind of be heard, but it sounds very choppy - this is what I had noticed about your early request ID (I can't say dropped packets were the issue, but the audio for that request was very choppy).

The best way to validate the audio is outlined here: https://github.com/deepgram/streaming-test-suite#validate-your-audio - essentially, if you fork your audio, you can then inspect it (I use Audacity haha), if the audio is choppy or distorted by the time it is sent to Deepgram, then we will produce terrible transcripts (and don't worry, this kind of issue comes up a lot, which is why we made that streaming test suite and validation guide). If that doesn't do the trick, I can try out your script on one of my servers and poke around, would have some time tomorrow for that.

Last question, just to cover the bases, can you share your TwiML Bin (minus whatever information in it might be sensitive)?

**Reply by dquang-tran** [reply]
> Can you share a request id at this point too? If the transcript is gibberish, it's likely the audio is still gibberish/corrupt/not formatted well for some reason, see the below point too:

Yeah! It's 2419dccc-5a23-4c24-bf6a-6564acfe06b1

> Twilio can drop packets, and when it does audio will come in with timestamps that seem to skip, like 0, 20, 40, 80, 100 (notice the 60-80ms audio was skipped)

Gotcha, copied your logic over to handle this case.

> Last question, just to cover the bases, can you share your TwiML Bin (minus whatever information in it might be sensitive)?

```
<?xml version="1.0" encoding="UTF-8"?>
<Response>
    <Start>
        <Stream url="wss://[NGROK_SERVER]/media" track="both_tracks"/>
     </Start>
     <Dial>[PHONE_NUMER]</Dial>
</Response>
```
^ Was partially following https://www.twilio.com/docs/voice/tutorials/consume-real-time-media-stream-using-websockets-python-and-flask



Will check out the test suite as well

**Reply by nikolawhallon** [reply]
Hold that thought, we're good. I checked `2419dccc-5a23-4c24-bf6a-6564acfe06b1` and it looks great, I get transcripts like:

```
one
two
three
four five
six
one two three
hello
one two
three four five
seven
eight
...
```

Your current issue is that you have:
```
      'multichannel': True,
      'channels': 2,
```
but you are only sending the `inbound` channel, so Deepgram is interpreting that single channel of audio as two channels, chopping it up, and probably producing terrible transcripts on the chopped up audio. Given you have `track="both_tracks"` in your TwiML Bin, I imagine you want to transcribe both channels - great - but before doing that, can you just turn `multichannel` off and set `'channels': 1` to make sure you can get to the point where you are getting transcripts like I pasted above?

**Reply by nikolawhallon** [reply]
(sending both `inbound` an `outbound` channels is tricky - Twilio streams them separately, but you need to stitch them together into valid stereo audio - the Twilio blog post goes over this, but I'd recommending trying that out after making sure you are getting valid transcripts for just the one channel first)

**Reply by dquang-tran** [reply]
That fixed it! Yeah, I initially had it multi channel because I eventually plan to take input from both, but was only feeding inbound for now for testing purposes. It's a bit counter-intuitive that it would result in such vastly different output. 

Really appreciate your help!

**Reply by dquang-tran** [reply]
Will be happy to contribute the code as sample when I have time — feels like this is a fairly common use case
